File: kotipalo_leo_ex03/p02/mdmorse/c/inout.c
/* $Id: inout.c 22 2008-11-12 13:06:32Z aakurone $ */

/* Input-output subroutines for the mdmorse code */

#include "global.h"

#define MAXBUF 200

void ReadParams(double         *initialT,
		double         *m,
		double         *tmax,
		double         *deltat,
		struct vector  *size,
		struct ivector *periodic,
		int            *nneighbourlistupdate,
		double         *rpotcut,
		double         *rskincut,  
		double         *morseD,
		double         *morsealpha,
		double         *morser0,
		int            *nmovieoutput,
		double         *btctau,
		double         *desiredT,
		double         *bpctau,
		double         *bpcbeta,
		double         *desiredP,
		int            *seed)


{
  int  i,n;
  FILE *fp;
  char  string[MAXBUF],buf[MAXBUF],*c;
  double x;
  
  fp=fopen("mdmorse.in","r");
   
  if (fp==NULL) { puts("Open error for file mdmorse.in"); exit(1); }

  /* Initialize variables to sensible (?) default values */
  *initialT=0.0; *m=63.546;
  size->x=0.0; size->y=0.0; size->z=0.0;

  *deltat=1.0;
  *nneighbourlistupdate=5;
  *rpotcut=4.02; *rskincut=5.0;
  /* Initialize Morse parameters for Cu */
  /* Values from Girifalco and Weizer, Phys. Rev. 114 (1959) 687. */
  *morseD=0.3429; *morsealpha=1.3588; *morser0=2.866;
  *btctau=0.0; *desiredT=0.0;
  *bpctau=0.0; *bpcbeta=1e-3; *desiredP=0.0;

  *seed=23267761;


  
  /* Loop through file and find all variables */
  i=0;
  while(1) {
     i=i+1;
     c=fgets(buf,MAXBUF,fp);
     if (c==NULL) break;

     string[0]=' ';
     n=sscanf(buf,"%s %lg",string,&x);

     /* If not parameter line, skip this line */
     if (string[0] != '$') continue;

     if (n<2) {
       printf("ERROR: Line \"%s\" looks like data line\n",buf);
       printf("but does not contain data in correct format\n");
       exit(0);
     }

     /* Look for variable-defining strings */
     if (strcmp(string,"$initialT")==0) *initialT=x;
     else if (strcmp(string,"$desiredT")==0) *desiredT=x;
     else if (strcmp(string,"$btctau")==0) *btctau=x;

     else if (strcmp(string,"$mass")==0) *m=x;

     else if (strcmp(string,"$bpctau")==0) *bpctau=x;
     else if (strcmp(string,"$bpcbeta")==0) *bpcbeta=x;
     else if (strcmp(string,"$desiredP")==0) *desiredP=x;

     else if (strcmp(string,"$xsize")==0) size->x=x;
     else if (strcmp(string,"$ysize")==0) size->y=x;
     else if (strcmp(string,"$zsize")==0) size->z=x;

     else if (strcmp(string,"$periodicx")==0) periodic->x=(int)(x+0.5);
     else if (strcmp(string,"$periodicy")==0) periodic->y=(int)(x+0.5);
     else if (strcmp(string,"$periodicz")==0) periodic->z=(int)(x+0.5);
     
     else if (strcmp(string,"$seed")==0) *seed=(int)(x+0.5);
     
     else if (strcmp(string,"$tmax")==0) *tmax=x;
     else if (strcmp(string,"$deltat")==0) *deltat=x;

     else if (strcmp(string,"$nupdate")==0) *nneighbourlistupdate=(int)(x+0.5);

     else if (strcmp(string,"$rpotcut")==0) *rpotcut=x;
     else if (strcmp(string,"$rskincut")==0) *rskincut=x;
     
     else if (strcmp(string,"$morseD")==0) *morseD=x;
     else if (strcmp(string,"$morsealpha")==0) *morsealpha=x;
     else if (strcmp(string,"$morser0")==0) *morser0=x;

     else if (strcmp(string,"$nmovieoutput")==0) *nmovieoutput=(int)(x+0.5);

     else {
       printf("Unknown parameter %s in input file\n",string);
       exit(0);
     }

     printf("Read in parameter %s to value %g\n",string,x);

  } /* End of readin loop */

  fclose(fp);

  /* Print out some general information about read results: */

  printf("%s %3d %3d %3d\n","Using periodics (1=on, 0=off)",
	 periodic->x,periodic->y,periodic->z);

  printf("%s %12.6g %12.6g %12.6g\n","Morse potential parameters: D alpha r0",
       *morseD,*morsealpha,*morser0);

  printf("%s %d %s\n","Movie output selected every",*nmovieoutput," steps");

  if (*btctau > 0.0) {
    printf("Doing Berendsen temperature control with tau T %g %g\n",
	   *btctau,*desiredT);
  }

  if (*bpctau > 0.0) {
    printf("Doing Berendsen pressure control with tau beta %g %g\n",
	   *bpctau,*bpcbeta);
  }

  return;
  
}



void ReadAtoms(struct vector *x,
	       char           atomname[MAXAT+1][5],
	       int           *N)

{
  char comment[MAXBUF],buf[MAXBUF];
  int  i,n;
  FILE *fp;

  fp=fopen("atoms.in","r");

  if (fp==NULL) { puts("File atoms.in open failed"); exit(0); }

  fgets(buf,MAXBUF,fp); sscanf(buf,"%d",N);

  if (*N > MAXAT) { puts("Error: MAXAT overflow, increase it."); exit(0); }

  fgets(comment,MAXBUF,fp);
  printf("%s%d%s\"%s\"\n","Reading in ",*N," atoms described as ",comment);
  
  for (i=1;i<= *N;i++) {

    fgets(buf,MAXBUF,fp); 
    n=sscanf(buf,"%s %lg %lg %lg",atomname[i],
	   &(x[i].x),&(x[i].y),&(x[i].z));

    if (n<4) {
	printf("Atom read in error for line \"%s\" n %d\n",buf,n);
	printf("%s %lg %lg %lg\n",atomname[i],x[i].x,x[i].y,x[i].z);
	puts("Atom read in error"); 
	exit(0);
    }

  }
  
  
}



void WriteAtoms(struct vector  *x,
		char            atomname[MAXAT+1][5],
		double         *Ekin,
		double         *Epot,
		int             N,
		double          time,
		struct vector   size,
		int             itime)
  
{
  
  char  comment[MAXBUF];
  int  i;
  static FILE *fp;

  static int firsttime=1;

  if (firsttime) {
    fp=fopen("atoms.out","w");
    if (fp==NULL) { puts("File atoms.out open failed"); exit(0); }
    firsttime=0;
  }

  fprintf(fp," %d\n",N);
  
  /*
  fprintf(fp,"%s %11.3f %s %12.4f %12.4f %12.4f\n",
	  "mdmorse atom output at time",
  	  time,"fs boxsize ",size.x,size.y,size.z);
  */
  fprintf(fp," Frame number %d %g fs boxsize %g %g %g\n",itime,time,size.x,size.y,size.z);
  
  for(i=1;i<=N;i++) {
    fprintf(fp,"%2s %12.6f%12.6f%12.6f%11.6f\n",
	    atomname[i],x[i].x,x[i].y,x[i].z,Epot[i]);
  }

  fflush(fp);
  
}

File: kotipalo_leo_ex03/p02/mdmorse/f90/inout.f90
! $Id: inout.f90 26 2009-07-23 12:11:52Z aakurone $

! Input-output subroutines for the mdmorse code

subroutine ReadParams(initialT,m,tmax,deltat,size,periodic,  &
     nneighbourlistupdate,rpotcut,rskincut,  &
     morseD,morsealpha,morser0,nmovieoutput, &
     btctau,desiredT,bpctau,bpcbeta,desiredP,seed,aoutt0,noutput)
     
  USE GLOBAL
  implicit none

  real(double)  :: initialT,m,tmax
  type(vector)  :: size
  type(ivector) :: periodic
  real(double)  :: deltat,rpotcut,rskincut,morseD,morsealpha,morser0
  integer       :: nneighbourlistupdate,nmovieoutput,seed,noutput
  real(double)  :: btctau,desiredT,bpctau,bpcbeta,desiredP,aoutt0
  
  integer :: i,ios
  character(80) :: line
  character(40) :: string
  real(double) :: x
  
  open(20,file='mdmorse.in',status='old',iostat=ios)
   
  if (ios /= 0) STOP 'Open error for file mdmorse.in'

  ! Initialize variables to sensible (?) default values
  initialT=0.0; m=63.546;
  size=vector(0.0,0.0,0.0)

  deltat=1.0;
  nneighbourlistupdate=5;
  rpotcut=4.02; rskincut=5.0;
  ! Initialize Morse parameters for Cu
  ! Values from Girifalco and Weizer, Phys. Rev. 114 (1959) 687.
  morseD=0.3429; morsealpha=1.3588; morser0=2.866;

  btctau=0.0; desiredT=0.0;
  bpctau=0.0; bpcbeta=1e-3; desiredP=0.0;
  
  seed=23267761
  vfromxyz=.false.
  scattering=.false.

  aoutt0=-1.0
  noutput=1

  
  ! Loop through file and find all variables
  i=0
  do
     i=i+1

     ! First read in line in whole.
     read(20,fmt='(A80)',iostat=ios) line
     ! Check for probable end-of-file
     if (ios < 0) exit

     ! If not parameter line, cycle
     if (line(1:1) /= '$') cycle
     
     ! Attempt to parse line into command and value using internal formatted read.
     read(unit=line,fmt=*,iostat=ios) string,x
     if (ios > 0) then
        print *,'ERROR: Data read in error on line',i
        stop 'Invalid parameter file'
     endif

     ! From here on line should be in variable format
     
     ! Look for variable-defining strings
     if (string=='$initialT') then
	initialT=x
     else if (string=='$desiredT') then
	desiredT=x
     else if (string=='$btctau') then
	btctau=x
     else if (string=='$mass') then
	m=x
     else if (string=='$xsize') then
	size%x=x
     else if (string=='$ysize') then
	size%y=x 
     else if (string=='$zsize') then
	size%z=x
     else if (string=='$periodicx') then
	periodic%x=int(x+0.5)
     else if (string=='$periodicy') then
	periodic%y=int(x+0.5)
     else if (string=='$periodicz') then
	 periodic%z=int(x+0.5)
     else if (string=='$seed') then
	seed=int(x+0.5)
     else if (string=='$tmax') then
	tmax=x
     else if (string=='$deltat') then
	deltat= x
     else if (string=='$nupdate') then
	nneighbourlistupdate=int(x+0.5)
     else if (string=='$rpotcut') then
	rpotcut=x 
     else if (string=='$rskincut') then
	rskincut=x     
     else if (string=='$morseD') then
	morseD=x
     else if (string=='$morsealpha') then
	morsealpha=x
     else if (string=='$morser0') then
	 morser0=x
     else if (string=='$bpctau') then
	 bpctau=x
     else if (string=='$bpcbeta') then
	 bpcbeta=x
     else if (string=='$desiredP') then
	 desiredP=x
     else if (string=='$aoutt0') then
	 aoutt0=x
     else if (string=='$noutput') then
	noutput=int(x+0.5)
     else if (string=='$nmovieoutput') then
	nmovieoutput=int(x+0.5)
     else if (string=='$scattering') then
	scattering=int(x)>0
     else
	print '(A,A)','Unknown parameter',string
	stop 'Parameter read in error'
     endif

     print '(A,A16,A,G13.6)','Read in parameter ',string,' value',x
    
  enddo

  if (initialT<0.0) then
     vfromxyz=.true.
     print '(a)','Atom velocities read from the xyz file.'
  end if
  if (scattering) then
     print '(a)','Scattering study.'
  end if

  close(20)

  ! Print out some general information about read results:

  print '(A,3I2)','Using periodics (1=on, 0=off)',     &
       periodic%x,periodic%y,periodic%z

  print '(A,3F12.6)','Morse potential parameters: D alpha r0', &
       morseD,morsealpha,morser0

  print '(A,I0,A)','Output selected every ',noutput,' steps'
  print '(A,I0,A)','Movie output selected every ',nmovieoutput,' steps'

  if (btctau > 0.0) then 
     print '(/,A,2F10.3,/)','Doing Berendsen temperature control with tau T', &
	  btctau,desiredT
  endif
  
  if (bpctau > 0.0) then 
     print '(/,A,2F10.3,/)','Doing Berendsen pressure control with tau beta', &
	  bpctau,bpcbeta
  endif
  
  return
  
end subroutine ReadParams


subroutine ReadAtoms(x,v,atomname,N)
  USE GLOBAL
  implicit none

  integer :: N
  type(vector) :: x(*),v(*)
  character(5) atomname(*)
  
  character(80) :: comment
  integer :: i,ios

  open(30,file='atoms.in',status='old',iostat=ios)

  if (ios/=0) STOP 'File atoms.in open failed'

  read(30,fmt=*) N

  if (N > MAXAT) STOP 'Error: MAXAT overflow, increase it.'

  read(30,fmt='(A80)') comment
  print '(A,I7,A,A40)','Reading in ',N,' atoms described as ',comment
  
  do i=1,N

     if (vfromxyz) then
        read(30,fmt=*,iostat=ios) atomname(i),x(i)%x,x(i)%y,x(i)%z,v(i)%x,v(i)%y,v(i)%z
     else
        read(30,fmt=*,iostat=ios) atomname(i),x(i)%x,x(i)%y,x(i)%z
     end if

     if (ios/=0) then
	print *,'Atom read in error for atom',i
	print *,atomname(i),x(i)%x,x(i)%y,x(i)%z
	STOP ' Atom read in error'
     endif

  enddo
  
  
end subroutine ReadAtoms



subroutine WriteAtoms(x,atomname,Ekin,Epot,N,time,size,Wat,istep)
  USE GLOBAL
  implicit none

  type(vector) :: x(*)
  character(5) :: atomname(*)
  integer :: N,istep
  real(double) :: Ekin(*),Epot(*),time,Wat(*)
  type(vector) :: size
  
  character(80) :: comment
  integer :: i,ios

  logical, save :: firsttime=.true.

  if (firsttime) then
     open(31,file='atoms.out',status='replace',iostat=ios)
     if (ios/=0) STOP 'File atoms.out open failed'
     firsttime=.false.
  endif
     

  write(31,fmt='(I10)') N

  !write(31,fmt='(A,F11.3,A,3F12.4)') 'mdmorse atom output at time ',time,' fs boxsize ',size%x,size%y,size%z
  write(31,fmt='('' Frame number '',i7,g14.5,'' fs boxsize'',3(F11.6,1X))') istep,time,size%x,size%y,size%z
  
  do i=1,N
     !write(31,fmt='(A2,1X,3g18.8,2g16.6)') atomname(i),x(i)%x,x(i)%y,x(i)%z,Epot(i),Ekin(i)
     write(31,fmt='(A2,1X,3g18.8,2g16.6)') atomname(i),x(i)%x,x(i)%y,x(i)%z,Epot(i),Wat(i)
  enddo
  
end subroutine WriteAtoms


subroutine scatteringoutput(time,x,v,Epot,Ekin)
  USE GLOBAL
  implicit none
  type(vector) :: x(*),v(*)
  real(double) :: time,Epot(*),Ekin(*)
  character(len=100),save :: form='(4g18.8)'

  write(100,fmt=form) time,x(1)%x,x(1)%y,x(1)%z
  write(200,fmt=form) time,x(2)%x,x(2)%y,x(2)%z

  write(101,fmt=form) time,v(1)%x,v(1)%y,v(1)%z
  write(201,fmt=form) time,v(2)%x,v(2)%y,v(2)%z

  write(102,fmt=form) time,Epot(1),Ekin(1)
  write(202,fmt=form) time,Epot(2),Ekin(2)

  return
end subroutine scatteringoutput
